### ISSUE TITLE
SubModelSelectedItem throw ArgumentNullException

### ISSUE THREAD

**Issue by eu-genes** [elmish/Elmish.WPF#371]
Hi, keep working with `4.0.0-beta-40`

repo: https://github.com/eu-genes/PerfSubModelSeqInSubModelOpt
`branche - selecteditem`
1. Run the program
2. Select any row
3. And sometimes  app throw ArgumentNullException
   - logger write `Setting selected VM to null`
   - System.ArgumentNullException with parametr name: item
   ```
   System.Windows.Automation.Peers.DataGridItemAutomationPeer..ctor(Object item, DataGridAutomationPeer dataGridPeer)
   System.Windows.Automation.Peers.DataGridAutomationPeer.CreateItemAutomationPeer(Object item)
   System.Windows.Automation.Peers.ItemsControlAutomationPeer.FindOrCreateItemAutomationPeer(Object item)
   System.Windows.Automation.Peers.DataGridAutomationPeer.RaiseAutomationSelectionEvents(SelectionChangedEventArgs e)
   System.Windows.Controls.DataGrid.OnSelectionChanged(SelectionChangedEventArgs e)
   System.Windows.Controls.Primitives.Selector.InvokeSelectionChanged(List`1 unselectedInfos, List`1 selectedInfos)
   System.Windows.Controls.Primitives.Selector.SelectionChanger.End()
   System.Windows.Controls.SelectedItemCollection.EndUpdateSelectedItems()
   System.Windows.Controls.Primitives.MultiSelector.EndUpdateSelectedItems()
   System.Windows.Controls.DataGrid.MakeFullRowSelection(ItemInfo info, Boolean allowsExtendSelect, Boolean allowsMinimalSelect)
   System.Windows.Controls.DataGrid.HandleSelectionForCellInput(DataGridCell cell, Boolean startDragging, Boolean allowsExtendSelect, Boolean allowsMinimalSelect)
   System.Windows.Controls.DataGridCell.OnAnyMouseLeftButtonDown(MouseButtonEventArgs e)
   System.Windows.Controls.DataGridCell.OnAnyMouseLeftButtonDownThunk(Object sender, MouseButtonEventArgs e)
   System.Windows.Input.MouseButtonEventArgs.InvokeEventHandler(Delegate genericHandler, Object genericTarget)
   System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   System.Windows.UIElement.ReRaiseEventAs(DependencyObject sender, RoutedEventArgs args, RoutedEvent newEvent)
   System.Windows.UIElement.OnMouseDownThunk(Object sender, MouseButtonEventArgs e)
   System.Windows.Input.MouseButtonEventArgs.InvokeEventHandler(Delegate genericHandler, Object genericTarget)
   System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   System.Windows.UIElement.RaiseTrustedEvent(RoutedEventArgs args)
   System.Windows.UIElement.RaiseEvent(RoutedEventArgs args, Boolean trusted)
   System.Windows.Input.InputManager.ProcessStagingArea()
   System.Windows.Input.InputManager.ProcessInput(InputEventArgs input)
   System.Windows.Input.InputProviderSite.ReportInput(InputReport inputReport)
   System.Windows.Interop.HwndMouseInputProvider.ReportInput(IntPtr hwnd, InputMode mode, Int32 timestamp, RawMouseActions actions, Int32 x, Int32 y, Int32 wheel)
   System.Windows.Interop.HwndMouseInputProvider.FilterMessage(IntPtr hwnd, WindowMessage msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   System.Windows.Interop.HwndSource.InputFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
   MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   MS.Win32.UnsafeNativeMethods.DispatchMessage(MSG& msg)
   System.Windows.Threading.Dispatcher.PushFrameImpl(DispatcherFrame frame)
   System.Windows.Threading.Dispatcher.PushFrame(DispatcherFrame frame)
   System.Windows.Application.RunDispatcher(Object ignore)
   System.Windows.Application.RunInternal(Window window)
   System.Windows.Application.Run(Window window)
   System.Windows.Application.Run()
   SubModelSeqInSubModelOpt.App.Main()
  ```

**Comment by TysonMN** [comment]
This is a really interesting problem.  I don't fully understand it yet.

First, I think this bad behavior was introduced in PR #359.  In particular, your code works correctly when
https://github.com/elmish/Elmish.WPF/blob/66cf95600404704c140a1174f49e9c8baec35730/src/Elmish.WPF/WpfProgram.fs#L70

is changed to
```fsharp
fun msg -> element.Dispatcher.Invoke(fun () -> innerDispatch msg)
```

However, this shouldn't be necessary.  Invoking either synchronously or asynchronously should work.

Even so, there is an interesting difference between these two approaches.  In the synchronous approach, `TryGetMember` is only called once
```
[main.UserModel] TrySetMember SelectedUser
New message: UsersMsg (Select (Some 1)) Updated state: ...
[main.UserModel] PropertyChanged SelectedUser
[main.UserModel] TryGetMember SelectedUser
```
but it is called twice in the asynchronous approach
```
[main.UserModel] TrySetMember SelectedUser
[main.UserModel] TryGetMember SelectedUser
New message: UsersMsg (Select (Some 1)) Updated state: ...
[main.UserModel] PropertyChanged SelectedUser
[main.UserModel] TryGetMember SelectedUser
```

Surprisingly, I obtained that last logging sequence by replacing
```xaml
<Menu Grid.Row="0">
    <MenuItem Header="SubModels" >
        <Button Content="UsersSubModel" Command="{Binding ShowUsers}"></Button>
    </MenuItem>
</Menu>
```
with
```xaml
<Button Content="UsersSubModel" Command="{Binding ShowUsers}"/>
```

This makes me think the real problem is bug in WPF, but I don't know yet for sure.

**Comment by TysonMN** [comment]
Another thing that makes me think this is a bug in WPF is that the [FirstChanceException](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.firstchanceexception) and [UnhandledException](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.unhandledexception) events are not raised by this exception.

Another odd thing is that the bug doesn't happen if the initial state of the model is the state that results after executing the command.  But the bug still happens after executing the command even after starting with that initial state.

**Comment by eu-genes** [comment]
Thanks for your research, unfortunately i'm newbie in WPF

**Comment by TysonMN** [comment]
Even changing from
```xaml
<Menu Grid.Row="0">
    <MenuItem Header="SubModels" >
        <MenuItem Content="UsersSubModel" Command="{Binding ShowUsers}"></Button>
    </MenuItem>
</Menu>
```
to
```xaml
<Menu Grid.Row="0">
    <MenuItem Content="UsersSubModel" Command="{Binding ShowUsers}"></Button>
</Menu>
```
is a workaround.

**Comment by TysonMN** [comment]
I am so confident that the fundamental issue is a bug in WPF that I submitted [this bug report](https://github.com/dotnet/wpf/issues/4279).

**Comment by TysonMN** [comment]
Fixed in [4.0.0-beta-41](https://www.nuget.org/packages/Elmish.WPF/4.0.0-beta-41).

**Comment by eu-genes** [comment]
> Fixed in [4.0.0-beta-41](https://www.nuget.org/packages/Elmish.WPF/4.0.0-beta-41).
Thanks, it works!
